### DISCUSSION TITLE
pop, push

### DISCUSSION THREAD

**Discussion by junwonjang** [remeda/remeda#960] [Suggest a Function]
Using `hasAtLeast`, I am able to validate the length of an array and avoid the need to optionally chain elements of the array, like so:
```
const array = [1,2,3]
assert.ok(hasAtLeast(array, 3));
array[2].toString(); // no need for optional chain
```

However, if you call `array.pop`, typescript will still think the array has length 3 even though it was mutated to have one less element. Here is an example:
```
const array = [1,2,3]
assert.ok(hasAtLeast(array, 3));
array.pop();
array[2].toString(); // ts thinks this array still has 3 things
```

This is likely a typescript issue and not a remeda issue, but hopefully this is helpful to at least document to help others avoid any confusing gotchas!

**Comment by eranhirsch** [comment]
Improving the default typing is one of the reasons Remeda offers our version for built-in functions like map, filter, etc...

The biggest limitation of how the JavaScript lib is built is that the functions are tacked onto the object prototypes (e.g. `Array.prototype`), which don't provide typescript enough information to infer the return type from the input type.

To work around the issue you describe, we can offer our own `pop` and `push` functions.

**Comment by cjquines** [comment]
for generality, we might want to instead offer `splice` (or `toSpliced`) which generalizes both pop and push

this would complement things like #470 (and perhaps allows us to deprecate `swapIndices`, and not add `moveIndex`)
